### DISCUSSION TITLE
[Proposal]: Allow member names to match their containing type name

### DISCUSSION THREAD

**Discussion by Youssef1313** [dotnet/csharplang#5763] [Language Ideas]
# Allow member names to match their containing type name

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Removes the restriction where the compiler produces an error if a member name matches its containing type name:

> error CS0542: '{0}': member names cannot be the same as their enclosing type

## Motivation
[motivation]: #motivation

There are use cases where a developer will want a member name to match its containing type name.

## Detailed design
[design]: #detailed-design

It's possible that we'll need to only remove this code:

https://github.com/dotnet/roslyn/blob/5193908c7283d377b9133a24e0e4394c603869c6/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs#L1272-L1290

which will allow the following to compile:

```csharp
public class C
{
    public int C;
}
```

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by Youssef1313** [comment]
FYI @jeffhandley

**Comment by theunrepentantgeek** [comment]
Simply removing the code triggering CS0542 isn't enough; you also need to address the underlying issue that the error is preventing.

As I recall, CS0542 prevents an ambiguity in the language, a situation where the language parser would have to do a lot of speculative work to try and resolve the ambiguity.

Unfortunately, my Google skills have failed me and I can't find a reference for this.

However, I did find an interesting example:

``` csharp
public class Person
{
    public string KnownAs { get; set; }
    public string FamilyName { get; set; }
    public string LegalName { get; set; }

    public void Person(string knownAs, string familyName, string legalName)
    {
        // ... obvious code elided ...
    }
}
```

In this case, the method `Person` _**is not a constructor**_ - the return type makes it a method. Without CS0542 this common beginner error wouldn't be directly detected by the compiler.

Without any explicit constructor, the compiler would synthesize a parameterless constructor, and the user would be left scratching their head as to why `new Person()` works, but `new Person(known, family, legal)` fails.

There has quite literally been an _enormous_ amount of work put into making the compiler errors of C# as intelligible and directly actionable as possible - a language change like this that provides little (if any) upside while introducing significant downsides for those unfamiliar with the language is not likely to get any traction.

And that's best case, on the assumption that my earlier recollection about underlying ambiguities in the language is incorrect.

**Reply by jnm2** [reply]
This could be addressed by a compiler warning on void-returning instance methods with the same name as the containing type.

**Comment by Youssef1313** [comment]
I don't know if this was the reason for the error, but I don't see this being confusing. It is not hard for beginners to spot an extra `void` (or return type) in my opinion.

**Reply by erik-kallen** [reply]
I have worked with C# for almost 20 years, and I still miss it. For the same reason that it is is hard to spot accidentally doubled word in a sentence.

**Comment by theunrepentantgeek** [comment]
If it's "not hard", explain the number of blog posts and Stack Overflow questions that are easily found - and those people have the compiler telling them exactly the problem. It gets much harder to spot the problem when the compiler starts being cryptic.

It's extremely easy, as adepts at C#, to gloss over the things that beginners find difficult.

**Comment by YairHalberstadt** [comment]
> It's possible that we'll need to only remove this code:

> https://github.com/dotnet/roslyn/blob/5193908c7283d377b9133a24e0e4394c603869c6/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs#L1272-L1290

I don't think it would be that simple, since you would want to disallow this for a nested type, which would almost always be ambiguous whether you're referring to the inner or outer type. There's probably other cases I'm not thinking of

**Comment by Youssef1313** [comment]
@YairHalberstadt We *could* pick up the nested type for that case (might not be good), or disallow it as you suggested (I think this is better).

**Comment by 333fred** [comment]
Unless there is an ldm champion, this needs to start as a discussion, not an issue. As such, I'm converting it over.

**Comment by TahirAhmadov** [comment]
I recall this question being asked back in the day and somebody, perhaps Eric Lippert, answering it in detail. In short, I think this is extremely difficult to accomplish, like @theunrepentantgeek wrote above. The team looked into this long and hard and couldn't make it work. This is a pain point for me sometimes but I've learned to live with it.

**Comment by BenjaminBrienen** [comment]
I would like if this proposal were accepted and implemented. One use case would be deserializing a schema that would otherwise require a JsonPropertyName applied to a "misnamed" property to properly support.

"Beginners would be confused about constructors" does not strike me as a valid reason to disallow this.

In terms of whether or not it's possible to have this, I imagine it's fairly similar to having a class with the same name as the namespace it's in. It's recommended against, but it's still possible to do at the moment. After all, a class is basically a namespace itself. CSharp also supports naming a variable/property the same as its type. Those two things make me think that having a member named the same as its containing type name should be possible to do, and should be allowed.
